File: jambeth/jambeth-cache/src/main/java/com/koch/ambeth/cache/util/PrefetchCommand.java
package com.koch.ambeth.cache.util;

/*-
 * #%L
 * jambeth-cache
 * %%
 * Copyright (C) 2017 Koch Softwaredevelopment
 * %%
 * Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

     http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
 * #L%
 */

import com.koch.ambeth.merge.util.DirectValueHolderRef;

public class PrefetchCommand
{
	public final DirectValueHolderRef valueHolder;

	public final PrefetchPath[] prefetchPaths;

	public PrefetchCommand(DirectValueHolderRef valueHolder, PrefetchPath[] cachePaths)
	{
		this.valueHolder = valueHolder;
		this.prefetchPaths = cachePaths;
	}

	@Override
	public int hashCode()
	{
		return valueHolder.hashCode() ^ prefetchPaths.hashCode();
	}

	@Override
	public boolean equals(Object obj)
	{
		if (this == obj)
		{
			return true;
		}
		if (!(obj instanceof PrefetchCommand))
		{
			return false;
		}
		PrefetchCommand other = (PrefetchCommand) obj;
		// Use equals() of ValueHolderKey
		return valueHolder.equals(other.valueHolder) && prefetchPaths == other.prefetchPaths;
	}
}

File: ambeth/Ambeth.Cache/ambeth/util/PrefetchCommand.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Reflection;
using De.Osthus.Ambeth.Log;
using De.Osthus.Ambeth.Util;
using De.Osthus.Ambeth.Transfer;
using De.Osthus.Ambeth.Merge.Transfer;
using System.Collections.ObjectModel;
using De.Osthus.Ambeth.Cache;
using System.Text;
using De.Osthus.Ambeth.Merge.Model;
using De.Osthus.Ambeth.Ioc;
using De.Osthus.Ambeth.Merge;
using De.Osthus.Ambeth.Typeinfo;

namespace De.Osthus.Ambeth.Util
{
    public class PrefetchCommand
    {
        public readonly DirectValueHolderRef valueHolder;

        public readonly PrefetchPath[] prefetchPaths;

        public PrefetchCommand(DirectValueHolderRef valueHolder, PrefetchPath[] prefetchPaths)
        {
            this.valueHolder = valueHolder;
            this.prefetchPaths = prefetchPaths;
        }

        public override int GetHashCode()
        {
            return valueHolder.GetHashCode() ^ prefetchPaths.GetHashCode();
        }

        public override bool Equals(object obj)
        {
            if (Object.ReferenceEquals(this, obj))
            {
                return true;
            }
            if (!(obj is PrefetchCommand))
            {
                return false;
            }
            PrefetchCommand other = (PrefetchCommand)obj;
            if (valueHolder is DirectValueHolderRef)
		    {
			    // Use equals() of ValueHolderKey
			    return valueHolder.Equals(other.valueHolder) && prefetchPaths == other.prefetchPaths;
		    }
            return Object.ReferenceEquals(valueHolder, other.valueHolder)
                && Object.ReferenceEquals(prefetchPaths, other.prefetchPaths);
        }
    }
}